ocaml/tests/test_repository_helpers.ml

```diff
@@ -2750,52 +2757,720 @@ module ParseUpdateInfoList = Generic.MakeStateless (struct
+        )
+      ; ([EvacuateHost; RestartDeviceModel; RebootHost], [RebootHost])
+      ; ([EvacuateHost; RestartToolstack], [RestartToolstack; EvacuateHost])
+      ; ([EvacuateHost; RestartToolstack; RebootHost], [RebootHost])
+      ; ([EvacuateHost; RebootHost], [RebootHost])
+      ; ([RestartDeviceModel], [RestartDeviceModel])
+      ; ( [RestartDeviceModel; RestartToolstack]
+        , [RestartToolstack; RestartDeviceModel]
+        )
+      ; ([RestartDeviceModel; RestartToolstack; RebootHost], [RebootHost])
+      ; ([RestartDeviceModel; RebootHost], [RebootHost])
+      ; ([RestartToolstack], [RestartToolstack])
+      ; ([RestartToolstack; RebootHost], [RebootHost])
+      ; ([RebootHost], [RebootHost])
+      ; ([RestartVM], [RestartVM])
+      ; ([RestartVM; EvacuateHost], [EvacuateHost; RestartVM])
+      ; ( [RestartVM; EvacuateHost; RestartDeviceModel]
+        , [EvacuateHost; RestartVM]
+        )
+      ; ( [RestartVM; EvacuateHost; RestartDeviceModel; RestartToolstack]
+        , [RestartToolstack; EvacuateHost; RestartVM]
+        )
+      ; ( [
+            RestartVM
+          ; EvacuateHost
+          ; RestartDeviceModel
+          ; RestartToolstack
+          ; RebootHost
+          ]
+        , [RebootHost; RestartVM]
+        )
+      ; ( [RestartVM; EvacuateHost; RestartDeviceModel; RebootHost]
+        , [RebootHost; RestartVM]
+        )
+      ; ( [RestartVM; EvacuateHost; RestartToolstack]
+        , [RestartToolstack; EvacuateHost; RestartVM]
+        )
+      ; ( [RestartVM; EvacuateHost; RestartToolstack; RebootHost]
+        , [RebootHost; RestartVM]
+        )
+      ; ([RestartVM; EvacuateHost; RebootHost], [RebootHost; RestartVM])
+      ; ([RestartVM; RestartDeviceModel], [RestartVM])
+      ; ( [RestartVM; RestartDeviceModel; RestartToolstack]
+        , [RestartToolstack; RestartVM]
+        )
+      ; ( [RestartVM; RestartDeviceModel; RestartToolstack; RebootHost]
+        , [RebootHost; RestartVM]
+        )
+      ; ([RestartVM; RestartDeviceModel; RebootHost], [RebootHost; RestartVM])
+      ; ([RestartVM; RestartToolstack], [RestartToolstack; RestartVM])
+      ; ([RestartVM; RestartToolstack; RebootHost], [RebootHost; RestartVM])
+      ; ([RestartVM; RebootHost], [RebootHost; RestartVM])
+      ]
+end)
+
+module GuidanceSetReduceTest = Generic.MakeStateless (struct
+  module Io = struct
+    type input_t = Guidance.t list * Guidance.t list
+
+    type output_t = Guidance.t list
+
+    let string_of_input_t (l1, l2) =
+      Fmt.(str "%a" Dump.(list string)) (List.map Guidance.to_string l1)
+      ^ " + "
+      ^ Fmt.(str "%a" Dump.(list string)) (List.map Guidance.to_string l2)
+
+    let string_of_output_t l =
+      Fmt.(str "%a" Dump.(list string)) (List.map Guidance.to_string l)
+  end
+
+  let transform (l1, l2) =
+    let open GuidanceSet in
+    reduce (of_list l1) (of_list l2) |> elements
+
+  let tests =
+    let open Guidance in
+    `QuickAndAutoDocumented
+      [
```

Comment:
I think the test cases exhausted all the combination possibilities, we'd better have a mechanism to ensure no one is missed later.